File: onebot/src/main/java/cn/evole/onebot/sdk/util/json/MessageEventAdapter.java
package cn.evole.onebot.sdk.util.json;


import cn.evole.onebot.sdk.entity.Anonymous;
import cn.evole.onebot.sdk.event.message.GroupMessageEvent;
import cn.evole.onebot.sdk.event.message.GuildMessageEvent;
import cn.evole.onebot.sdk.event.message.MessageEvent;
import cn.evole.onebot.sdk.event.message.PrivateMessageEvent;
import cn.evole.onebot.sdk.util.JsonHelper;
import com.google.gson.*;

import java.lang.reflect.Type;

public class MessageEventAdapter extends JsonHelper implements JsonDeserializer<MessageEvent> {
    public MessageEvent deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) {
        MessageEvent e = null;
        JsonObject obj = json.getAsJsonObject();
        String postType = obj.get("post_type").getAsString();
        String subType = obj.get("sub_type").getAsString();
        long time = obj.get("time").getAsLong();
        long selfId = obj.get("self_id").getAsLong();
        String messageType = obj.get("message_type").getAsString();
        long userId = obj.get("user_id").getAsLong();
        String message = forceString(obj, "message");
        String rawMessage = ignorable(obj, "raw_message", "");
        int font = ignorable(obj, "font", 0);

        switch (messageType) {
            case "group": {
                e = groupMessageEvent(obj, subType);
                break;
            }
            case "private": {
                e = privateMessageEvent(obj, subType);
                break;
            }
            case "guild": { // go-cqhttp, OpenShamrock
                e = guildMessageEvent(obj, subType);
                break;
            }
        }
        if (e != null) {
            e.setPostType(postType);
            e.setTime(time);
            e.setSelfId(selfId);
            e.setMessageType(messageType);
            e.setUserId(userId);
            e.setMessage(message);
            e.setRawMessage(rawMessage);
            e.setFont(font);
        }
        return e;
    }

    private MessageEvent groupMessageEvent(JsonObject obj, String subType) {
        int messageId = MsgAdapter.getMessageId(obj);
        long groupId = obj.get("group_id").getAsLong();
        Anonymous anonymous = fromJson(obj, "anonymous", Anonymous.class);
        GroupMessageEvent.GroupSender sender = fromJson(obj, "sender", GroupMessageEvent.GroupSender.class);
        return new GroupMessageEvent(messageId, subType, groupId, anonymous, sender);
    }

    private MessageEvent privateMessageEvent(JsonObject obj, String subType) {
        int messageId = MsgAdapter.getMessageId(obj);
        int tempSource = ignorable(obj, "temp_source", Integer.MIN_VALUE); // OpenShamrock
        PrivateMessageEvent.PrivateSender sender = fromJson(obj, "sender", PrivateMessageEvent.PrivateSender.class);
        long groupId = ignorable(obj, "group_id", sender != null ? sender.groupId : 0); // OpenShamrock
        String fromNick = ignorable(obj, "from_nick", sender != null ? sender.nickname : ""); // OpenShamrock
        return new PrivateMessageEvent(messageId, subType, sender, groupId, tempSource, fromNick);
    }

    private MessageEvent guildMessageEvent(JsonObject obj, String subType) {
        String messageId = obj.get("message_id").getAsString();
        String guildId = obj.get("guild_id").getAsString();
        String channelId = obj.get("channel_id").getAsString();
        String selfTinyId = obj.get("self_tiny_id").getAsString();
        GuildMessageEvent.GuildSender sender = fromJson(obj, "sender", GuildMessageEvent.GuildSender.class);
        return new GuildMessageEvent(messageId, subType, guildId, channelId, selfTinyId, sender);
    }

    public static class PrivateSender implements JsonDeserializer<PrivateMessageEvent.PrivateSender> {
        @Override
        public PrivateMessageEvent.PrivateSender deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
            JsonObject obj = json.getAsJsonObject();
            PrivateMessageEvent.PrivateSender sender = new PrivateMessageEvent.PrivateSender();
            sender.userId = obj.get("user_id").getAsLong();
            sender.groupId = ignorable(obj, "group_id", 0);
            sender.nickname = obj.get("nickname").getAsString();
            sender.sex = ignorable(obj, "sex", "unknown");
            sender.age = ignorable(obj, "age", 0);
            return sender;
        }
    }
}

File: overflow-core/src/main/kotlin/cn/evolvefield/onebot/sdk/util/json/MessageEventAdapter.kt
package cn.evolvefield.onebot.sdk.util.json

import cn.evolvefield.onebot.sdk.entity.PrivateSender
import cn.evolvefield.onebot.sdk.event.message.GroupMessageEvent
import cn.evolvefield.onebot.sdk.event.message.GuildMessageEvent
import cn.evolvefield.onebot.sdk.event.message.MessageEvent
import cn.evolvefield.onebot.sdk.event.message.PrivateMessageEvent
import cn.evolvefield.onebot.sdk.util.*
import cn.evolvefield.onebot.sdk.util.json.MsgAdapter.Companion.messageId
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import java.lang.reflect.Type

class MessageEventAdapter : JsonDeserializerKt<MessageEvent> {
    override fun deserializeFromJson(
        json: JsonElement,
        type: Type,
        ctx: JsonDeserializationContext
    ): MessageEvent? {
        val obj = json.asJsonObject
        val subType = obj.string("sub_type")
        val messageType = obj.string("message_type")
        return when (messageType) {
            "group" -> obj.groupMessage(subType)
            "private" -> obj.privateMessage(subType)
            "guild" -> obj.guildMessage(subType)
            else -> null
        }?.apply {
            postType = obj.string("post_type")
            time = obj.long("time")
            selfId = obj.long("self_id")
            userId = obj.long("user_id")
            isJsonMessage = obj.has("message") && obj["message"].isJsonArray
            message = obj.forceString("message")
            rawMessage = obj.ignorable("raw_message", "")
            font = obj.ignorable("font", 0)
        }
    }

    private fun JsonObject.groupMessage(subType: String): MessageEvent = GroupMessageEvent().apply {
        this.subType = subType
        messageId = messageId()
        groupId = long("group_id")
        anonymous = fromJson("anonymous")
        sender = fromJson("sender")
    }
    private fun JsonObject.privateMessage(subType: String): MessageEvent = PrivateMessageEvent().apply {
        this.subType = subType
        messageId = messageId()
        tempSource = ignorable("temp_source", Int.MIN_VALUE)
        sender = fromJson("sender")!!
        groupId = ignorable("group_id", sender.groupId)
        fromNick = ignorable("from_nick", sender.nickname)
    }
    private fun JsonObject.guildMessage(subType: String): MessageEvent = GuildMessageEvent().apply {
        this.subType = subType
        messageId = string("message_id")
        guildId = string("guild_id")
        channelId = string("channel_id")
        selfTinyId = string("self_tiny_id")
        sender = fromJson("sender")!!
    }

    class PrivateSenderAdapter : JsonDeserializer<PrivateSender> {
        override fun deserialize(
            json: JsonElement,
            typeOfT: Type,
            context: JsonDeserializationContext
        ): PrivateSender {
            val obj = json.asJsonObject
            return PrivateSender().apply {
                userId = obj["user_id"].asLong
                groupId = obj.ignorable("group_id", 0L)
                nickname = obj["nickname"].asString
                sex = obj.ignorable("sex", "unknown")
                age = obj.ignorable("age", 0)
            }
        }
    }
}
